### PR TITLE
DEP-8 integration testing

### PR DESCRIPTION
**Pull request by lene**
DEP-8 CI separated out. Depends on https://github.com/Nicotine-Plus/nicotine-plus/pull/106 to actually do anything.

### CODE REVIEW

File: debian/tests/control [hunk 1]
```diff
@@ -1,10 +1,31 @@
...
 #
 
 Test-Command: nicotine --version | grep Nicotine+
 Depends: nicotine
 Restrictions: isolation-container
 Features: test-name=test-version
 
+Test-Command: pytest-3 tests/unit
+Depends:
+    nicotine,
```
Comment (kiplingw) [inline debian/tests/control:13]: You can replace this line with @ per [spec](https://salsa.debian.org/ci-team/autopkgtest/raw/master/doc/README.package-tests.rst).

File: debian/tests/control [hunk 2]
```diff
@@ -1,10 +1,31 @@
...
+    python3-pytest
+Restrictions: isolation-container
+Features: test-name=unit-tests
+
+Test-Command: xvfb python3 tests/integration/run_app.py
+Depends:
+    nicotine,
```
Comment (kiplingw) [inline debian/tests/control:20]: Same thing.

File: debian/tests/control [hunk 3]
```diff
@@ -1,10 +1,31 @@
...
+    python3-dogtail,
+    python3-pyatspi,
+    python3-cairo,
```
Comment (kiplingw) [inline debian/tests/control:23]: Are all of these explicit dependencies actually necessary for the test? You can safely remove any that `nicotine` itself depends on.

File: tests/unit/data/login/socket_localhost:22420.log [hunk 4]
```diff
@@ -0,0 +1,86 @@
+(dp0
```
Comment (kiplingw) [inline tests/unit/data/login/socket_localhost:22420.log:1]: This whole file and any other derived files shouldn't be checked into SCM.

File: debian/tests/control (hunk 1 above)
Comment (lene) [reply debian/tests/control]: Alright, done.

File: debian/tests/control (hunk 2 above)
Comment (lene) [reply debian/tests/control]: Same.

File: debian/tests/control (hunk 3 above)
Comment (lene) [reply debian/tests/control]: Yes, this list can be much shorter actually.

File: tests/unit/data/login/socket_localhost:22420.log (hunk 4 above)
Comment (lene) [reply tests/unit/data/login/socket_localhost:22420.log:1]: This is needed to mock the server traffic in unit tests. The ones in the `data/search` folder are not utilized yet though, so I took them out.

**Comment by kiplingw** [comment]
Hey @lene. It looks like that wonderful and comprehensive CI test you wrote wasn't actually working. Although it was likely already failing within the container, the exit code you were checking against was ${true} which presumably is non-zero, when what we really wanted to check for was nicotine(1) exiting with exit status zero.

I've fixed the false negative, but your test itself needs to be fixed. It looks like it is something to do with `gi.repository.Notify` not being correctly imported. It could also be that the DBus session bus isn't running and needs to be spawned prior to starting the test.

**Comment by lene** [comment]
Hi @kiplingw, not sure what you are talking about. The test is not checking the exit code, but the PID of the nicotine process - signifying just that it has started and hasn't crashed after 3 seconds - for which `${True}` is the correct value to check against. 
Not sure I understand you correctly, but looking through https://github.com/Nicotine-Plus/nicotine-plus/commit/958d4b5b6bba429ec735876cd9651f0e9e614168 makes me think you have maybe misinterpreted that?

**Comment by kiplingw** [comment]
Hey @lene. You are probably right. However, the test is still failing. This is what I'm seeing. Are you able to replicate?

```
...
autopkgtest [10:16:29]: test test-installed-artifacts.sh: [---------------------
*** Installing test harness...
Collecting robotframework
  Downloading robotframework-3.2.1-py2.py3-none-any.whl (618 kB)
Installing collected packages: robotframework
Successfully installed robotframework-3.2.1
*** Launching Nicotine+ inside of dummy frame buffer...
==============================================================================
Nicotine :: Basic integration tests for nicotine+ Start nicotine both as a ...
==============================================================================
Run nicotine as shell command                                         
(nicotine:3030): dbind-WARNING **: 10:16:33.643: AT-SPI: Error retrieving access
Traceback (most recent call last):
  File "/usr/bin/nicotine", line 276, in <module>
    run()
  File "/usr/bin/nicotine", line 252, in run
    app = frame.MainApp(data_dir, config, plugins, trayicon, hidden, bindip, po)
  File "/usr/lib/python3/dist-packages/pynicotine/gtkgui/frame.py", line 3729, _
    self.frame = NicotineFrame(data_dir, config, plugins, trayicon, start_hidde)
  File "/usr/lib/python3/dist-packages/pynicotine/gtkgui/frame.py", line 240, i_
    gi.require_version('Notify', '0.7')
  File "/usr/lib/python3/dist-packages/gi/__init__.py", line 129, in require_ven
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace Notify not available
| FAIL |
AssertionError
------------------------------------------------------------------------------
Nicotine :: Basic integration tests for nicotine+ Start nicotine b... | FAIL |
1 critical test, 0 passed, 1 failed
1 test total, 0 passed, 1 failed
==============================================================================
Output:  /tmp/autopkgtest.7vlBrU/build.16f/src/output.xml
Log:     /tmp/autopkgtest.7vlBrU/build.16f/src/log.html
Report:  /tmp/autopkgtest.7vlBrU/build.16f/src/report.html
autopkgtest [10:16:38]: test test-installed-artifacts.sh: ---------------------]
autopkgtest [10:16:39]: test test-installed-artifacts.sh:  - - - - - - - - - - -
test-installed-artifacts.sh FAIL non-zero exit status 1
autopkgtest [10:16:39]:  - - - - - - - - - - running shell - - - - - - - - - -
...
```

**Comment by lene** [comment]
Yes I could. It is an example of CI doing its job actually (I must say I'm getting pretty convinced Debian CI is useful! :-D) Created https://github.com/Nicotine-Plus/nicotine-plus/pull/196 to fix.

**Comment by kiplingw** [comment]
Other than some minor Gtk warnings, it looks like it works!

```
autopkgtest [11:27:24]: test test-installed-artifacts.sh: [---------------------
*** Installing test harness...
Collecting robotframework
  Downloading robotframework-3.2.1-py2.py3-none-any.whl (618 kB)
Installing collected packages: robotframework
Successfully installed robotframework-3.2.1
*** Launching Nicotine+ inside of dummy frame buffer...
==============================================================================
Nicotine :: Basic integration tests for nicotine+ Start nicotine both as a ...
==============================================================================
Run nicotine as shell command                                         
(nicotine:3042): dbind-WARNING **: 11:27:27.998: AT-SPI: Error retrieving access

(nicotine:3042): Gtk-WARNING **: 11:27:28.324: Error loading theme icon 'list-ru

(nicotine:3042): Gtk-WARNING **: 11:27:28.325: Error loading theme icon 'edit-c 

(nicotine:3042): Gtk-WARNING **: 11:27:28.325: Error loading theme icon 'edit-c 

(nicotine:3042): Gtk-WARNING **: 11:27:28.325: Error loading theme icon 'edit-r 

(nicotine:3042): Gtk-WARNING **: 11:27:28.326: Error loading theme icon 'procesu

(nicotine:3042): Gtk-WARNING **: 11:27:28.327: Error loading theme icon 'list-ru

(nicotine:3042): Gtk-WARNING **: 11:27:28.327: Error loading theme icon 'edit-c 

(nicotine:3042): Gtk-WARNING **: 11:27:28.327: Error loading theme icon 'edit-c 

(nicotine:3042): Gtk-WARNING **: 11:27:28.327: Error loading theme icon 'procesu

(nicotine:3042): Gtk-WARNING **: 11:27:28.328: Error loading theme icon 'edit-c 

(nicotine:3042): Gtk-WARNING **: 11:27:28.329: Error loading theme icon 'documeu

(nicotine:3042): Gtk-WARNING **: 11:27:28.329: Error loading theme icon 'view-ru

(nicotine:3042): Gtk-WARNING **: 11:27:28.329: Error loading theme icon 'view-ru

(nicotine:3042): Gtk-WARNING **: 11:27:28.329: Error loading theme icon 'view-ru

(nicotine:3042): Gtk-WARNING **: 11:27:28.329: Error loading theme icon 'list-au

(nicotine:3042): Gtk-WARNING **: 11:27:28.329: Error loading theme icon 'list-au

(nicotine:3042): Gtk-WARNING **: 11:27:29.052: Error loading theme icon 'list-ru

(nicotine:3042): Gtk-WARNING **: 11:27:29.053: Error loading theme icon 'list-ru
| PASS |
------------------------------------------------------------------------------
Nicotine :: Basic integration tests for nicotine+ Start nicotine b... | PASS |
1 critical test, 1 passed, 0 failed
1 test total, 1 passed, 0 failed
==============================================================================
Output:  /tmp/autopkgtest.RVcdvd/build.7gg/src/output.xml
Log:     /tmp/autopkgtest.RVcdvd/build.7gg/src/log.html
Report:  /tmp/autopkgtest.RVcdvd/build.7gg/src/report.html
*** OK!
autopkgtest [11:27:32]: test test-installed-artifacts.sh: ---------------------]
autopkgtest [11:27:33]: test test-installed-artifacts.sh:  - - - - - - - - - - -
test-installed-artifacts.sh PASS
autopkgtest [11:27:34]: @@@@@@@@@@@@@@@@@@@@ summary
test-version         PASS (superficial)
unit-tests           PASS
test-installed-artifacts.sh PASS
```

I have a feeling the test may not have been running at all before, or it may have been but its output appeared to have been suppressed.

**Comment by kiplingw** [comment]
@mathiascode, could you please update the autopkgtest workflow so that it performs testing on all of our currently supported releases? Those are `Xenial`, `Bionic`, `Eoan`, `Focal`, and `Groovy`.

**Comment by mathiascode** [comment]
Github Actions has images for Xenial, Bionic and Focal.

**Pull request merged**